File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.cc
//11725
#include <iostream>
#include <vector>

using namespace std;

vector<vector<int>> a(100005);
int parent[100005];

void dfs(int x) {
    for (int i = 0; i < a[x].size(); i++) {
        int y = a[x][i];
        if (parent[y] == 0) {
            parent[y] = x;
            dfs(y);
        }
    }
}

int main() {
    int n; cin >> n;

    for (int i = 0; i < n-1; i++) {
        int u, v;
        cin >> u >> v;

        a[u].push_back(v);
        a[v].push_back(u);
    }

    parent[1] = 1;
    dfs(1);

    for (int i = 2; i <= n; i++) {
        cout << parent[i] << "\n";
    }
    
    return 0;
}

File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.py
#11725
import sys
from collections import deque

sys.setrecursionlimit(10**6)

def input():
    return sys.stdin.readline().rstrip()

n = int(input())  # 'node' 대신 'n'으로 변경
graph = [[] for _ in range(n+1)]
parent = [0] * (n+1)
parent[1] = -1  # root

for _ in range(n-1):
    u, v = map(int, input().split())
    graph[u].append(v)
    graph[v].append(u)

def dfs(v):
    for neighbor in graph[v]:
        if parent[neighbor] == 0:
            parent[neighbor] = v
            dfs(neighbor)

dfs(1)

for i in range(2, n+1):
    print(parent[i])